[MM-9516] Add feature to add user to channel after at-mention

saturninoabril opened the pull request:
#### Summary
Add feature to add user to channel after at-mention

#### Ticket Link
Jira ticket: [MM-9516](https://mattermost.atlassian.net/browse/MM-9516)

#### Checklist
- [x] Has change on redux https://github.com/mattermost/mattermost-redux/pull/426
- [x] Has UI changes

#### Device Information
This PR was tested on: [iPhone 8, iOS 11, Android]

#### Screenshots

### At-mentioned  a user at Channel view of public channel
<img width="332" alt="screen shot 2018-03-17 at 1 57 59 am" src="https://user-images.githubusercontent.com/5334504/37547620-c68e4fae-29ad-11e8-9a43-5f495b0bbbce.png">

### At-mentioned a user at Thread screen of private channel
![screen shot 2018-03-19 at 7 04 28 pm](https://user-images.githubusercontent.com/5334504/37592111-abd22968-2ba8-11e8-9fa5-62e37c08bb9f.png)

### Added a user at private channel (ephemeral post is posted on Thread screen)
![screen shot 2018-03-19 at 7 04 01 pm](https://user-images.githubusercontent.com/5334504/37592112-abfe1ce4-2ba8-11e8-93fb-44fa1de001b5.png)

### At-mentioned multiple users at Thread screen of private channel
![screen shot 2018-03-19 at 7 05 28 pm](https://user-images.githubusercontent.com/5334504/37592110-aba17c28-2ba8-11e8-8f37-f37d16110796.png)

### Added one and multiple users at private channel (ephemeral posts are posted on Thread screen)
![screen shot 2018-03-19 at 7 05 46 pm](https://user-images.githubusercontent.com/5334504/37592109-ab4e98c8-2ba8-11e8-8586-9d268a49a41e.png)

enahum commented on app/components/post_add_channel_member/index.js:
```diff
@@ -0,0 +1,81 @@
+function sendEphemeralPost(user, addedUsername, message, channelId, postRootId) {
+    const timestamp = Date.now();
+
+    const post = {
```
I think all this has to be inside the thunk

saturninoabril commented:
Hmmm, please hold on review. I still have one issue to resolve. I thought I have resolved it already.

enahum commented on app/components/post_add_channel_member/post_add_channel_member.js:
```diff
@@ -0,0 +1,196 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import {
+    defineMessages,
```
This is not needed, you can just define objects with id and defaultMessage

enahum commented on app/components/post_add_channel_member/post_add_channel_member.js:
```diff
@@ -0,0 +1,196 @@
+            sendEphemeralPost: PropTypes.func.isRequired,
+        }).isRequired,
+        currentUser: PropTypes.object.isRequired,
+        channelType: PropTypes.string.isRequired,
```
In the index you are passing this as empty in one of the cases so either change that or remove the required here

saturninoabril replied:
You're right. I'll just remove required here.

enahum commented on app/components/post_body/index.js:
```diff
@@ -26,16 +41,45 @@ function mapStateToProps(state, ownProps) {
+
+    let addMemberProps = {};
+    if (post.props && post.props.add_channel_member) {
+        addMemberProps = post.props.add_channel_member;
```
We are already passing the post props, do we really need this here?

saturninoabril replied:
You're right, this is no longer needed.

saturninoabril commented:
Thanks @enahum for your review. I have updated this based on your comments and also updated mattermost-redux commit to latest.
This is now ready for review.

enahum commented on app/components/post_add_channel_member/index.js:
```diff
@@ -0,0 +1,80 @@
+    };
+}
+
+function sendEphemeralPost(user, addedUsername, message, channelId, postRootId) {
```
Can you move this to be an action under `app/actions/views` ?

saturninoabril replied:
Sure. I should have add test to `sendAddToChannelEphemeralPost` but I'm having issue with the mock store. Let me do that separately (next time) if that's fine with you.

enahum commented:
@saturninoabril rebase required

saturninoabril commented:
@enahum done rebase/fix merge conflicts, and did clean up

hmhealey commented on app/components/post_body/index.js:
```diff
@@ -26,6 +41,25 @@ function mapStateToProps(state, ownProps) {
+
+    let isPostAddChannelMember = false;
+    if (
+        (channel.type === General.PRIVATE_CHANNEL || channel.type === General.OPEN_CHANNEL) &&
```
I don't understand all this logic. Why can't we just check the post type?

saturninoabril replied:
The post type here is `system_ephemeral` which is commonly used.
I'll try to rearrange the logic which will only do things if necessary. Also, I realized that some checks were already covered by `canManageMembers`.

hmhealey replied:
Oh, I see. We should really separate the notion of ephemeral posts from post type at some point so that we can give a type to ephemeral posts. Could we just look at the post props then and skip the other checks? This message shouldn't be created for non-private/public channels anyway

hmhealey commented on app/selectors/post_list.js:
```diff
@@ -42,6 +43,10 @@ export function makePreparePostIdsForPostList() {
             for (let i = posts.length - 1; i >= 0; i--) {
                 const post = posts[i];
 
+                if (post.type === Posts.POST_TYPES.EPHEMERAL_ADD_TO_CHANNEL && !selectedPostId) {
```
I don't understand this. Why are we not showing EPHEMERAL_ADD_TO_CHANNEL posts when there's no selected post ID?

saturninoabril replied:
Maybe I should add a comment on this logic (let me know if needed).
The way to identify active screen is: Thread screen is active if `selectedPostId` is not empty, and then any other screen like Channel if it's empty.
When we added a user at Thread screen, we are posting ephemeral post that the user is added on Thread screen itself. However we are not posting that same ephemeral post at Channel screen since there is a similar system post coming from the server that the user is added. If it's not filtered out, there will be 2 posts at Channel screen.

hmhealey replied:
Maybe we could change it on the server then so that it replies to the thread if you're at mentioning someone in a reply? Having the message sometimes created by the server and sometimes by the client seems error prone
